pkg/agent/controller/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,663 @@
+	c.enqueuePacketCapture(pc)
+}
+
+func nameToPath(name string) string {
+	return filepath.Join(packetDirectory, name+".pcapng")
+}
+
+func (c *Controller) worker() {
+	for c.processPacketCaptureItem() {
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		klog.ErrorS(err, "Error syncing PacketCapture, exiting", "key", key)
+	}
+	return true
+}
+
+func (c *Controller) cleanupPacketCapture(pcName string) {
+	path := nameToPath(pcName)
+	exist, err := afero.Exists(defaultFS, path)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if path exists", "path", path)
+	}
+	if !exist {
+		return
+	}
+	if err := defaultFS.Remove(path); err == nil {
+		klog.V(2).InfoS("Deleted pcap file", "name", pcName, "path", path)
+	} else {
+		klog.ErrorS(err, "Failed to delete pcap file", "name", pcName, "path", path)
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, nil, fmt.Errorf("packet file already exists. this may be due to an unexpected termination")
+	} else if os.IsNotExist(err) {
+		file, err = defaultFS.Create(filePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice trying to locate the target device for capture. If the target pod is not exist on the current node,
+// means this node will not perform the capture.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(pc *crdv1alpha1.PacketCapture) error {
+	var err error
+	defer func() {
+		if err != nil {
+			c.cleanupPacketCapture(pc.Name)
+			err := c.updatePacketCaptureStatus(pc, crdv1alpha1.PacketCaptureFailed, fmt.Sprintf("Node: %s, Error: %+v", c.nodeConfig.Name, err), 0)
+			if err != nil {
+				klog.ErrorS(err, "failed to update PacketCapture status")
+			}
+		}
+	}()
+
+	pcState := &packetCaptureState{name: pc.Name}
+	device := c.getTargetCaptureDevice(pc)
+	if device == nil {
+		return nil
+	}
+
+	packetSource, err := c.newPacketSourceFn(c, pc)
```

Comment:
You didn't handle the case when the return packetSource is 'nil'. I suppose you should return immediately when the packetSource is nil because of no device is found?

Reply:
Yes，you are right. Weird it didn't panic when it's nil.